MRG, VIZ, MAINT: Use new figure class for epochs.plot and ICA.plot_sources

drammock opened the pull request:
this is an element of #7751 

Basically working already, though needs testing and maybe some cosmetics.

hoechenberger commented:
Hello @drammock, maybe I'm doing something wrong – and I really haven't looked at the code changes here – but right-clicking on a channel name in a plot produced by `ICA.plot_sources(evoked)` should bring up a property plot, is that right? Right-clicking doesn't do anything for me on macOS, both with the default macOS and with the `Qt5Agg` backends. I surely must be missing something?

drammock commented:
at the moment only `ica.plot_sources(raw)` has been updated.  `ica.plot_sources(epochs)` and ``ica.plot_sources(evoked)` haven't yet been changed.

hoechenberger commented:
Thanks @drammock! It's beautiful!

What I've noticed:

- clicking on the Help button doesn't open the help window – once I right-click on a component name and the property plot opens, the help window opens behind the property window
- I get the macOS beach ball for a few seconds after doing the right click. I wonder if it would be possible to prevent this blocking, and if we could even print `processing …` or something similar to inform the user that we haven't crashed, but are merely busy?

drammock commented:
@jasmainak is it easy for you to test the autoreject package with this branch, to make sure that `epoch_colors` still works as expected?

drammock commented:
OK, now `raw.plot()`, `epochs.plot()`, `ica.plot_sources(raw)`, and `ica.plot_sources(epochs)` all use the new figure classes.  Assuming I didn't break anything with these last few commints, this is ready for review.  (sorry for the monstrous diff; it turned out to be easier to do ICA and epochs at the same time).  Notes:

1. here are the circle renderings of the tutorials I updated this time: [ica tutorial](https://23291-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/plot_40_artifact_correction_ica.html) and [epochs tutorial](https://23291-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/epochs/plot_20_visualize_epochs.html#plotting-epochs-as-time-series). Note that in the ICA tutorial I used `show_scrollbars=False` in all the `plot_sources()` calls, which I think is generally better for tutorials except that you lose the x-axis label.
2. bad ICA components now get grey color instead of red color, similar to bad channels in the raw and epoch plots.
3. I changed the keyboard shortcuts for scrolling through epochs in the time domain, to make them more consistent with `raw.plot()`.  Previously, in `epochs.plot()` → and ← would scroll a full window width at a time.  Now, I've set it so that → and ← scroll a single epoch, and now shift→ and shift← scroll a full window width at a time.  This is consistent with `raw.plot()` where → does 1/4 window width and shift→ does full window width.
4. The labeling of epochs at the bottom of the main window and along the bottom edge of the scrollbar are now *numbers from epochs.selection* instead of sequential indices. This overrides the work in #7299 so I'm keen to hear @SophieHerbst's opinion whether this is a welcome or unwelcome change.
5. I have deprecated the parameter `event_colors` in `epochs.plot()` in favor of the similarly-named `event_color` parameter. It turns out that `raw.plot()` has an `event_color` parameter that was generally more flexible than what used to be implemented for epochs. Now, they both have the same name, they work the same way, and they both do everything each of them used to be able to do. Here is the new docstring: https://github.com/mne-tools/mne-python/blob/cc4dff991ae5bf645bae32d130db88e77c98b780/mne/utils/docs.py#L98-L105
6. The horizontal scrollbar in the epochs plot now has divisions showing epoch boundaries.
7. clicking in the data axes of epochs plots is now more similar to raw plots. In particular, you must click on data traces to mark bad epochs (previously clicking between the data traces was good enough). Now, clicking between data traces places the green vertical line (like it does on raw plots) and right-clicking removes the green vertical line (again, like raw plots).
8. all three plot types allow right-click on channel names to bring up additional context plots. raw plots → channel location, epoch plots → epochs image for that channel, ica plots → a `plot_properties` window for that component.

of these, I think items 3, 4, 5, and 7 are **prominent (non-cosmetic) changes** which I think should get buy-in from at least a couple other devs before merge, and some of them probably deserve changelog entries.  Anything else you notice that seems weird  --- anything at all --- please mention it now.

larsoner commented:
> used show_scrollbars=False in all the plot_sources() calls, which I think is generally better for tutorials except that you lose the x-axis label.

We could allow `show_scrollbars='xaxis'` or something to specify just to keep the x axis. It would be better if the name were `decorate=True` then `decorate='xaxis'` would make more sense but I'm okay with a bit of a misnomer here (not worth deprecation cycle to make the name more general).

> This overrides the work in #7299

This appeared to just be about 0- versus 1-based indexing so I think it's fine / complementary

Otherwise all these API changes sound like improvements to me, +1 for all of them. I'll wait to do a thorough review until we see if others agree and no other major changes are needed.

@hoechenberger I see a heart, does that mean you're happy with all points?

hoechenberger commented:
> @hoechenberger I see a heart, does that mean you're happy with all points?

Yes I’m quite happy! Still got to actually try it out, but I do love the list of changes

drammock commented:
> We could allow `show_scrollbars='xaxis'` or something to specify just to keep the x axis. It would be better if the name were `decorate=True` then `decorate='xaxis'` would make more sense but I'm okay with a bit of a misnomer here (not worth deprecation cycle to make the name more general).

WDYT about just always showing an x-axis label, regardless of whether scrollbars are visible or hidden?

larsoner commented:
I'm not sure how that interacts with `zen` mode, but I never really use this so I'm happy to defer to your judgment on this

larsoner commented:
I have no strong opinion, up to you. I would say if you want to go with as much backward compat as possible add `show_scrollbars='xaxis'` option, if it's not a big concern then just change the behavior of `False`

agramfort reviewed (approved):
it works like a charm !!!! and with 2000 lines removed 🍾 

@drammock you are like 🎅 to me !

hoechenberger commented:
LOL I can imagine that scene!!
"Daddy, what's for Christmas?" – "Best present EVER! 2,000 SLOC removed, BOOYAA!!! 🤡" – "😱"

larsoner commented:
And longstanding functionality added (such as butterfly mode for epochs, right?) and waaaay more DRY code, etc. Bravo!

I'll take a look and try things Monday. @hoechenberger do you want to test it out? I don't use the epochs or ICA very much so I'm not the best person to do it.

larsoner commented:
Does this close #4850, or is more needed for that?

hoechenberger commented:
> @hoechenberger do you want to test it out? I don't use the epochs or ICA very much so I'm not the best person to do it.

Yes but won't be able to do it before tonight or tomorrow UTC+1

drammock commented:
> Does this close #4850, or is more needed for that?

No but it should be quite easy to do. Will add that.

drammock commented on mne/viz/raw.py:
```diff
@@ -552,13 +552,16 @@ def _setup_channel_selections(raw, kind, order):
     else:  # kind == 'selection'
         from ..io import RawFIF, RawArray
-        if not isinstance(raw, (RawFIF, RawArray)):
+        if isinstance(raw, BaseEpochs):
+            pass
+        elif not isinstance(raw, (RawFIF, RawArray)):
```
@larsoner is there a clever way to detect whether an Epochs object contains Neuromag data?  I was never satisfied for Raw checking against RawFIF and RawArray, so if you have a better idea on that too I'd love to hear it.

larsoner replied:
If this has to do with `selection` mode, I assume you are talking about Neuromag MEG data (as EEG can be treated like any other format). The robust way would be to check for coil_type of any of info['chs'] being one of the Neuromag types. Part of this sort of functionality might already live in find_layout somehow, because it needs to inspect channels to see if we have MEG channels that we can use a nice built-in layout for.

drammock commented:
Yep, it's for selection mode (which now works for epochs). I'll `git grep` for coil_type and see what I can find. Thanks.

Sent from ProtonMail mobile

-------- Original Message --------
On Nov 7, 2020, 18:53, Eric Larson wrote:

> @larsoner commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [mne/viz/raw.py](https://github.com/mne-tools/mne-python/pull/8381#discussion_r519240306):
>
>> +        if isinstance(raw, BaseEpochs):
> +            pass
> +        elif not isinstance(raw, (RawFIF, RawArray)):
>
> If this has to do with selection mode, I assume you are talking about Neuromag MEG data (as EEG can be treated like any other format). The robust way would be to check for coil_type of any of info['chs'] being one of the Neuromag types. Part of this sort of functionality might already live in find_layout somehow, because it needs to inspect channels to see if we have MEG channels that we can use a nice built-in layout for.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/mne-tools/mne-python/pull/8381#discussion_r519240306), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAN2AU4NYOC7O7TXO54T46LSOXT2LANCNFSM4STYZHFQ).

drammock commented:
Ready from my end. @larsoner I think this now cleanly implements `group_by='selection'` for epochs.  @hoechenberger have you had a chance to test this out yet?  Here's a quick testing script:

```python
import os
import numpy as np
import matplotlib.pyplot as plt
import mne
plt.ion()
sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False)
raw.del_proj()
raw.crop(tmax=20)
raw.apply_proj()
ecg_proj_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                             'sample_audvis_ecg-proj.fif')
ecg_projs = mne.read_proj(ecg_proj_file)
raw.add_proj(ecg_projs)
sample_data_noise_cov_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                          'ernoise-cov.fif')
noise_cov = mne.read_cov(sample_data_noise_cov_file)
events = mne.find_events(raw, stim_channel='STI 014')
event_dict = {'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3}
event_color = {1: 'r', 2: 'g', -1: 'y'}
epochs = mne.Epochs(raw, events, event_id=event_dict, preload=True, proj=False)
fig = epochs.plot(n_epochs=6, event_color=event_color, events=epochs.events,
                  event_id=event_dict, group_by='selection')
# for a treat (not useful, but needed to support autoreject package):
# epoch_colors = [np.random.choice(list('rgbcmyk'), epochs.info['nchan']
#                                  ).tolist() for _ in range(len(epochs))]
# fig = epochs.plot(epoch_colors=epoch_colors, butterfly=True)
```

hoechenberger commented:
> Here's a quick testing script:

How convenient 😍 Will rest momentarily.

hoechenberger commented:
> Will rest momentarily.

Test. Will **test** momentarily. Ouch, Freudian 😅

hoechenberger commented:
Hey @drammock,

I'm currently forced to use my old 2013 MacBook Air, which has a small screen and is rather slow, and when trying to resize the Raw browser, I can reproducibly make Python crash. The traceback would like something like the following:

```python
Traceback (most recent call last):
  File "/tmp/foo.py", line 24, in <module>
    fig = epochs.plot(n_epochs=6, event_color=event_color, events=epochs.events,
  File "/Users/richard/Development/mne-python/mne/epochs.py", line 1111, in plot
    return plot_epochs(self, picks=picks, scalings=scalings,
  File "/Users/richard/Development/mne-python/mne/viz/epochs.py", line 881, in plot_epochs
    fig = _browse_figure(**params)
  File "/Users/richard/Development/mne-python/mne/viz/_figure.py", line 2112, in _browse_figure
    fig = _figure(inst=inst, toolbar=False, FigureClass=MNEBrowseFigure,
  File "/Users/richard/Development/mne-python/mne/viz/_figure.py", line 2102, in _figure
    fig = figure(FigureClass=FigureClass, **kwargs)
  File "/Users/richard/miniconda3/envs/mne/lib/python3.8/site-packages/matplotlib/pyplot.py", line 687, in figure
    figManager = new_figure_manager(num, figsize=figsize,
  File "/Users/richard/miniconda3/envs/mne/lib/python3.8/site-packages/matplotlib/pyplot.py", line 315, in new_figure_manager
    return _backend_mod.new_figure_manager(*args, **kwargs)
  File "/Users/richard/miniconda3/envs/mne/lib/python3.8/site-packages/matplotlib/backend_bases.py", line 3493, in new_figure_manager
    fig = fig_cls(*args, **kwargs)
  File "/Users/richard/Development/mne-python/mne/viz/_figure.py", line 357, in __init__
    self.subplotpars.update(left=left, bottom=bottom, top=top, right=right)
  File "/Users/richard/miniconda3/envs/mne/lib/python3.8/site-packages/matplotlib/figure.py", line 205, in update
    raise ValueError('bottom cannot be >= top')
ValueError: bottom cannot be >= top
```

Here's a video:
https://youtu.be/c_ddbYydzgs

I'm using `Qt5Agg`. Not sure if there's anything you can do about this…

I even once wasn't able to open the Raw browser anymore even after restarting Python, it would instantly crash with the above error message. After deleting `~/.mne`, it started working again.

Will look at the new features now and report back

drammock commented:
sounds like you hit #8304

hoechenberger commented:
@drammock I'm now constantly running into this, cannot really test your new viz :( Never saw this issue before, but this is also the smallest screen size I've ever used with MNE: 1366 x 768 pixels

hoechenberger commented:
@drammock I will try to give it another shot tomorrow, going to bed now :) But if you feel ready to get this merged, don't wait for me!

hoechenberger commented:
Thanks for addressing #8304, @drammock! This seems to be fixed for me now. However, bad news is, instead of this traceback plus (sometimes) a segfault I now **only** receive the segfault 💩Could be an issue with PyQt5 on macOS, I don't know. The problem still seems to be related to resizing somehow. I'm wondering if it would be possible to disable redrawing while the window is being resized? You know, like inn the Windows 3.11 days? Because I have the suspicion that this might be (part of) the issue here...

hoechenberger commented:
But I don't intend to filibuster this PR just because my old computer is having some troubles. If I'm the only one experiencing this, maybe we can just go ahead and merge for now… I unfortunately won't be able to provide feedback / test, though. But we can always iterate later. (Will hopefully have my 2019 MBP again in a week or two)

larsoner commented:
> Could be an issue with PyQt5 on macOS, I don't know.

Exceptions thrown during a callback will cause PyQt to abort. This should in theory be handled by `matplotlib` -- IIRC they have some mechanisms in place to make it so that any PyQt event handler of theirs does not pass on the exceptions (this is why they get printed instead). So this is probably a bug with their implementation somewhere. It would be great if we could isolate this and open an issue, or fix the bug in their code.

@hoechenberger are you on the latest matplotlib version? And what version of PyQt5?

larsoner commented on mne/viz/tests/test_raw.py:
```diff
@@ -273,11 +260,11 @@ def test_plot_raw_selection(raw, new_mpl):
+    assert sel_fig.lasso.ec[:, 0].sum() == 0   # all channels black
     # test lasso
     sel_fig._set_custom_selection()  # lasso empty → should do nothing
     sensor_ax = sel_fig.mne.sensor_ax
```
Okay now it looks like you can fix the test below:

https://travis-ci.org/github/mne-tools/mne-python/jobs/742563175#L6462-L6469

```
mne/viz/tests/test_raw.py:280: in test_plot_raw_selection
    assert got == want
E   AssertionError: assert ['MEG 0121', 'MEG 0122', 'MEG 0123'] == ['MEG 0111', 'MEG 0112', 'MEG 0131', 'MEG 0132', 'MEG 0133']
E     At index 0 diff: 'MEG 0121' != 'MEG 0111'
E     Right contains 2 more items, first extra item: 'MEG 0132'
E     Full diff:
E     - ['MEG 0111', 'MEG 0112', 'MEG 0131', 'MEG 0132', 'MEG 0133']
E     + ['MEG 0121', 'MEG 0122', 'MEG 0123']
```

larsoner commented on mne/viz/tests/test_raw.py:
```diff
@@ -308,7 +295,7 @@ def test_plot_raw_ssp_interaction(raw):
     # open SSP window
     _fake_click(fig, fig.mne.ax_proj, [0.5, 0.5])
-    assert len(plt.get_fignums()) == 2
+    assert len(plt.get_fignums()) == 2  # XXX this fails with minimal deps
```
revert, doesn't seem to fail?

drammock replied:
I had deleted that once, must have snuck back in during one of the many rebases

larsoner commented:
Pushed a commit to showcase `butterfly=True, group_by='selection'`. Feel free to remove this commit and put it elsewhere if the spot I chose isn't good enough. Other than that and the Travis failure, LGTM!

hoechenberger commented:
@larsoner 
I'm using Matplotlib 3.3.2, Qt 5.12.9, and PyQt 5.12.3

drammock commented:
@larsoner all green here, can we get this in so more folks can start testing it?

larsoner commented:
Yes, thanks @drammock !

agramfort commented:
awesome @drammock <https://github.com/drammock> !

>

hoechenberger commented:
🎈

larsoner commented:
@drammock looks like these are yours:

https://app.circleci.com/pipelines/github/mne-tools/mne-python/4732/workflows/ffa530e7-f767-4af5-b1b3-ebe0b679c0f9/jobs/23476

Can you check and open a quick PR to fix?
